File: compiler/src/main/java/com/raizlabs/android/dbflow/processor/definition/MigrationDefinition.java
package com.raizlabs.android.dbflow.processor.definition;

import com.raizlabs.android.dbflow.annotation.Migration;
import com.raizlabs.android.dbflow.processor.DBFlowProcessor;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;
import com.raizlabs.android.dbflow.processor.writer.FlowWriter;
import com.squareup.javawriter.JavaWriter;

import javax.lang.model.element.TypeElement;
import java.io.IOException;

/**
 * Description: Used in holding data about migration files.
 */
public class MigrationDefinition extends BaseDefinition implements FlowWriter {


    public String databaseName;

    public Integer version;

    public MigrationDefinition(ProcessorManager processorManager, TypeElement typeElement) {
        super(typeElement, processorManager);
        setDefinitionClassName("");

        Migration migration = typeElement.getAnnotation(Migration.class);

        this.databaseName = migration.databaseName();
        if(databaseName == null || databaseName.isEmpty()) {
            databaseName = DBFlowProcessor.DEFAULT_DB_NAME;
        }

        version = migration.version();
    }

    @Override
    public void write(JavaWriter javaWriter) throws IOException {

    }
}

File: processor/src/main/kotlin/com/dbflow5/processor/definition/MigrationDefinition.kt
package com.dbflow5.processor.definition

import com.dbflow5.annotation.Migration
import com.dbflow5.processor.ProcessorManager
import com.dbflow5.processor.utils.extractTypeNameFromAnnotation
import com.dbflow5.processor.utils.isNullOrEmpty
import com.grosner.kpoet.typeName
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.TypeElement

/**
 * Description: Used in holding data about migration files.
 */
class MigrationDefinition(
    migration: Migration,
    processorManager: ProcessorManager, typeElement: TypeElement
) : BaseDefinition(typeElement, processorManager) {

    var databaseName: TypeName?
    val version: Int
    val priority: Int

    var constructorName: String? = null
        private set

    init {
        setOutputClassName("")
        databaseName = migration.extractTypeNameFromAnnotation { it.database }
        version = migration.version
        priority = migration.priority

        val elements = typeElement.enclosedElements
        elements.forEach { element ->
            if (element is ExecutableElement && element.simpleName.toString() == "<init>") {
                if (!constructorName.isNullOrEmpty()) {
                    manager.logError(
                        MigrationDefinition::class,
                        "Migrations cannot have more than one constructor. " +
                            "They can only have an Empty() or single-parameter constructor Empty(Empty.class) that specifies " +
                            "the .class of this migration class."
                    )
                }

                if (element.parameters.isEmpty()) {
                    constructorName = "()"
                } else if (element.parameters.size == 1) {
                    val params = element.parameters
                    val param = params[0]

                    val type = param.asType().typeName
                    if (type is ParameterizedTypeName && type.rawType == ClassName.get(Class::class.java)) {
                        val containedType = type.typeArguments[0]
                        constructorName = CodeBlock.of("(\$T.class)", containedType).toString()
                    } else {
                        manager.logError(
                            MigrationDefinition::class,
                            "Wrong parameter type found for $typeElement. Found $type but required ModelClass.class"
                        )
                    }
                }
            }
        }
    }

}
